oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java

```diff
@@ -0,0 +1,197 @@
+                                .getChildNode(IndexDefinition.STATUS_NODE)
+                                .getProperty(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP).getValue(Type.DATE);
+                        long reindexCompletionTimeInMillis = PurgeOldVersionUtils.getMillisFromString(reindexCompletionTime);
+                        long currentTimeInMillis = System.currentTimeMillis();
+                        if (currentTimeInMillis - reindexCompletionTimeInMillis > thresholdInMillis) {
+                            activeProductVersion = indexNameObjectList.get(i).getProductVersion();
+                        }
+                    } else {
+                        LOG.warn(IndexDefinition.REINDEX_COMPLETION_TIMESTAMP
+                                + " property is not set for index " + indexNameObjectList.get(i).getNodeName());
+                    }
+                }
+            } else {
+            /*
+            Once we found active index, we mark all its custom version with DELETE
+            and OOTB same product version index with DELETE_HIDDEN_AND_DISABLE
+             */
+                if (indexNameObjectList.get(i).getProductVersion() == activeProductVersion
+                        && indexNameObjectList.get(i).getCustomerVersion() == 0) {
+                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObjectList.get(i));
+                    indexVersionOperation.setOperation(Operation.DELETE_HIDDEN_AND_DISABLE);
+                    indexVersionOperationList.add(indexVersionOperation);
+                } else {
+                    IndexVersionOperation indexVersionOperation = new IndexVersionOperation(indexNameObjectList.get(i));
+                    indexVersionOperation.setOperation(Operation.DELETE);
+                    indexVersionOperationList.add(indexVersionOperation);
+                }
+            }
+        }
+        if (!isValidIndexVersionOperationList(indexVersionOperationList)) {
+            indexVersionOperationList = Collections.emptyList();
+        }
+        return indexVersionOperationList;
+    }
+
+    /*
+        returns IndexNameObjects in descending order of version i.e from latest to oldest
+    */
+    private static List<IndexName> getSortedVersionedIndexNameObjects(List<IndexName> indexNameObjectList) {
+        indexNameObjectList.sort(new Comparator<IndexName>() {
+            @Override
+            public int compare(IndexName indexNameObj1, IndexName indexNameObj2) {
+                return indexNameObj2.compareTo(indexNameObj1);
+            }
+        });
+        return indexNameObjectList;
+    }
+
+    /**
+     * @param indexVersionOperations
+     * @return true if the IndexVersionOperation list passes following criteria.
+     * For merging indexes we need baseIndex and latest custom index.
+     * So we first validate that if there are custom indexes than OOTB index with same product must be marked with DELETE_HIDDEN_AND_DISABLE
```

Comment:
This is a quite important function, and I think it is correct... But would it make sense to add documentation, or even better test cases, with examples of what is valid and what is invalid, and why it's valid / invalid? If test case, the method would no longer be private, which I think is fine (no need to make it public... default access is fine).